core/commands/corenet.go

```diff
@@ -0,0 +1,520 @@
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = checkEnabled(n)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if !n.OnlineMode() {
+			res.SetError(errNotOnline, cmds.ErrClient)
+			return
+		}
+
+		proto := "/app/" + req.Arguments()[0]
+		if checkProtoExists(n.PeerHost.Mux().Protocols(), proto) {
+			res.SetError(errors.New("protocol handler already registered"), cmds.ErrNormal)
+			return
+		}
+
+		addr, err := ma.NewMultiaddr(req.Arguments()[1])
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		listener, err := cnet.Listen(n, proto)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		app := corenet.AppInfo{
+			Identity: n.Identity,
+			Protocol: proto,
+			Address:  addr,
+			Closer:   listener,
+			Running:  true,
+			Registry: &n.Corenet.Apps,
+		}
+
+		go acceptStreams(n, &app, listener)
+
+		n.Corenet.Apps.Register(&app)
+
+		// Successful response.
+		res.SetOutput(&CorenetAppInfoOutput{
+			Protocol: proto,
+			Address:  addr.String(),
+		})
+	},
+}
+
+func checkProtoExists(protos []string, proto string) bool {
+	for _, p := range protos {
+		if p != proto {
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+func acceptStreams(n *core.IpfsNode, app *corenet.AppInfo, listener cnet.Listener) {
+	for app.Running {
+		remote, err := listener.Accept()
+		if err != nil {
+			listener.Close()
+			break
+		}
+
+		local, err := manet.Dial(app.Address)
+		if err != nil {
+			remote.Close()
+			continue
+		}
+
+		stream := corenet.StreamInfo{
+			Protocol: app.Protocol,
+
+			LocalPeer: app.Identity,
+			LocalAddr: app.Address,
+
+			RemotePeer: remote.Conn().RemotePeer(),
+			RemoteAddr: remote.Conn().RemoteMultiaddr(),
+
+			Local:  local,
+			Remote: remote,
+
+			Registry: &n.Corenet.Streams,
+		}
+
+		n.Corenet.Streams.Register(&stream)
+		startStreaming(&stream)
+	}
+	n.Corenet.Apps.Deregister(app.Protocol)
+}
+
+func startStreaming(stream *corenet.StreamInfo) {
+	go func() {
+		io.Copy(stream.Local, stream.Remote)
+		stream.Close()
+	}()
+
+	go func() {
+		io.Copy(stream.Remote, stream.Local)
+		stream.Close()
+	}()
+}
+
+var corenetDialCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Dial to an application service.",
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("Peer", true, false, "Remote peer to connect to"),
+		cmds.StringArg("Protocol", true, false, "Protocol identifier."),
+		cmds.StringArg("BindAddress", false, false, "Address to listen for application/s (default: /ip4/127.0.0.1/tcp/0)."),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = checkEnabled(n)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if !n.OnlineMode() {
+			res.SetError(errNotOnline, cmds.ErrClient)
+			return
+		}
+
+		addr, peer, err := ParsePeerParam(req.Arguments()[0])
```

Comment:
It should be possible to connect only with PeerID, is that possible with this code?